### PR TITLE
Added featues to binary distribution

### PR DESCRIPTION
**Pull request by chitrak7**
The code is working on linux now.
@angeman @alpmestan @snowleopard

### CODE REVIEW

File: src/Rules/BinaryDist.hs [hunk 1]
```diff
@@ -61,6 +63,27 @@ bindistRules = do
...
+      writeFile' wrapperHaddockPath haddockWrapper
+
+    root -/- "bindist" -/- "ghc-*" -/- "wrappers/common" %> \wrapperCommonPath ->
+      writeFile' wrapperCommonPath commonWrapper
+
+    root -/- "bindist" -/- "ghc-*" -/- "wrappers/hsc2hs" %> \wrapperHsc2hsPath ->
+      writeFile' wrapperHsc2hsPath hsc2hsWrapper
+
+    root -/- "bindist" -/- "ghc-*" -/- "wrappers/runhaskell" %> \wrapperRunHaskellPath ->
+      writeFile' wrapperRunHaskellPath runhaskellWrapper
```
Comment (alpmestan) [inline src/Rules/BinaryDist.hs]: I think you could substantially shorten this bit of code by having an function to abstract over the general pattern:

``` haskell
  bindistWrapperRule root "ghc" wrapperGhcPath ghcWrapper
  [...]
  bindistWrapperRule root "hsc2hs" wrapperHsc2hsPath hsc2hsWrapper
  bindistWrapperRule root "runhaskell" wrapperRunHaskellPath runhaskellWrapper
```

File: src/Rules/BinaryDist.hs [hunk 2]
```diff
@@ -83,7 +106,7 @@ bindistInstallFiles :: [FilePath]
 bindistInstallFiles =
   [ "config.sub", "config.guess", "install-sh"
   , "mk" -/- "config.mk.in", "mk" -/- "install.mk.in"
-  , "settings.in", "README", "INSTALL"
+  , "mk" -/- "project.mk.in", "settings.in", "README", "INSTALL"
```
Comment (alpmestan) [inline src/Rules/BinaryDist.hs]: I don't see any occurence of `project.mk` or `project.mk.in` besides this one, does the code as it is now somehow implicitly requires it? Or something will, in an upcoming patch? Or is it just a leftover from an experiment of yours?

Comment (chitrak7) [reply src/Rules/BinaryDist.hs]: project.mk is a dependency of config.mk. I am using config.mk.

**Comment by chitrak7** [comment]
I have made the necessary changes @alpmestan @angerman @snowleopard an tested it on linux. 
Also, we are not shipping ghci, should I create a separate PR to do so?

**Comment by chitrak7** [comment]
@alpmestan Should we ship c header files as well. It is probably required by hsc2hs. If yes, then I can create a separate PR.

**Comment by alpmestan** [comment]
We do need to ship some C header files yes. This is what my bindist of 8.4.3 comes with:

``` sh
$ ls /nix/store/a5z79ybg660ymjdq8mh4mcr7p6cbpxg4-ghc-8.4.3-with-packages/lib/ghc-8.4.3/include/
Cmm.h               HsFFI.h     Rts.h     Stg.h  ffitarget.h    ghcconfig.h    ghcversion.h  stg/
DerivedConstants.h  MachDeps.h  RtsAPI.h  ffi.h  ghcautoconf.h  ghcplatform.h  rts/
```

**Comment by chitrak7** [comment]
Hi @alpmestan, I have figured out how to ship includes. But since this PR is already pretty big, I think I should send a separate PR once this is merged.

File: src/Rules/BinaryDist.hs (hunk 2 above)
Comment (alpmestan) [reply src/Rules/BinaryDist.hs]: OK, as long as it doesn't pull anything else with it, that should be fine :-)

**Comment by alpmestan** [comment]
@chitrak7 As you prefer. Most of your PR (in terms of # of lines changed) is about your Makefile changes, so it's still quite easily reviewable, so you can feel free to use this PR too unless @snowleopard objects.

File: src/Rules/BinaryDist.hs [hunk 3]
```diff
@@ -106,16 +112,174 @@ bindistMakefile = unlines
...
+  , "\t$(INSTALL_DIR) \"$(bindir)\""
+  , "\t$(foreach p, $(BINARY_NAMES),\\"
+  , "\t\t$(call installscript,$p,$(bindir)/$p,$(bindir),$(GHCBINDIR),$(GHCBINDIR)/$p,$(libdir),$(docdir),$(includedir)))"
+  , ""
+  , "# END INSTALL"
+  , "# -----------------------------------------------------------------------------"
   ]
+
+basename :: FilePath -> String
+basename path = last $ splitOn "/" path
```
Comment (snowleopard) [inline src/Rules/BinaryDist.hs]: Should this be replaced with `takeFileName`?

File: src/Rules/BinaryDist.hs [hunk 4]
```diff
@@ -106,16 +112,174 @@ bindistMakefile = unlines
...
+
+wrapper :: String -> String
```
Comment (snowleopard) [inline src/Rules/BinaryDist.hs]: Should the first argument be given type `FilePath` instead? It's not just any `String`, it's a file name.

File: src/Rules/BinaryDist.hs (hunk 3 above)
Comment (snowleopard) [reply src/Rules/BinaryDist.hs]: I mean the standard function:

https://hackage.haskell.org/package/filepath-1.4.2/docs/System-FilePath-Posix.html#v:takeFileName

File: src/Rules/BinaryDist.hs (hunk 4 above)
Comment (chitrak7) [reply src/Rules/BinaryDist.hs]: @snowleopard It is not exactly a filename. It is the name of the program we are installing program for.

File: src/Rules/BinaryDist.hs (hunk 3 above)
Comment (chitrak7) [reply src/Rules/BinaryDist.hs]: Resolved. Thanks @snowleopard

File: src/Rules/BinaryDist.hs (hunk 4 above)
Comment (snowleopard) [reply src/Rules/BinaryDist.hs]: Yes, but `takeFileName` returns a `FilePath`, not a `String`. I know it's just a synonym, but still it is worth following standard convention, where file/program names are given type `FilePath`.

File: src/Rules/BinaryDist.hs (hunk 2 above)
Comment (chitrak7) [reply src/Rules/BinaryDist.hs]: Yes, project.mk is independent.

File: src/Rules/BinaryDist.hs (hunk 4 above)
Comment (chitrak7) [reply src/Rules/BinaryDist.hs]: @snowleopard Resolved

**Comment by snowleopard** [comment]
@chitrak7 Thanks, looks good to me.

@alpmestan Any further comments from you?

**Comment by alpmestan** [comment]
No, looks good to me.

**Comment by snowleopard** [comment]
Thanks, merged!

**Pull request merged**